### ISSUE TITLE
Inferior performance of EMA model

### ISSUE THREAD

**Issue by syorami** [microsoft/Swin-Transformer#111]
I was trying to train swin transformer on a object detection task and used EMA mechanism to stabilize the training process. But I found that the EMA model of swin-t  had inferior performance compared with original model. I am wondering why this could happen as the EMA performance is much better with other conv nets. Here is the code of how I implement EMA:

```
def build_ema(config):
    model = build_model(config)
    for param in model.parameters():
        param.detach_()
    return model

for ema_v, model_v in zip(ema_model.state_dict().values(), model.module.state_dict().values()):
    ema_v.copy_(ema_v * 0.999 + (1 - 0.999) * model_v)
```

**Comment by zeliu98** [comment]
Hi, we have also tried using EMA in ImageNet classification. It doesn't enhance performance either. I‘m not quite clear about the reason, maybe the stability of the model itself is better.

**Comment by syorami** [comment]
> Hi, we have also tried using EMA in ImageNet classification. It doesn't enhance performance either. I‘m not quite clear about the reason, maybe the stability of the model itself is better.

I've found out that it's due to the `relative_position_index` buffer in the attention block, which is formatted in integer. After multiplied by the EMA ratio, the `relative_position_index` would be transferred back to integer and this would lead to the value change of indexes. The inferior performance comes with mismatched features. Just skip the EMA computation for `relative_position_index` works well for me.

**Comment by yxchng** [comment]
@syorami how much improvement do you get after your proposed fix?

**Comment by syorami** [comment]
> @syorami how much improvement do you get after your proposed fix?

This depends on your task and model capacity. Sry I didn't remember the actual numbers.

**Comment by BestJuly** [comment]
Hi @syorami, which task do you obtain improvements? For EMA updates, if updating `model.parameters()` instead of `model.state_dict()`, `relative_position_index` will not be involved as it does not require any gradients during training.

I tested SwinT on ImageNet, and have not observe any improvements yet, which is similar to the finding from @zeliu98 .

**Comment by ctxqlxs** [comment]
any update?
